Add Python free path to inverse_cancellation

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a Python free path to the inverse cancellation transpiler pass. Previously, the only way to run the pass required Python because the pass handles arbitrary gates to cancel which can include custom gates from Python. This meant to validate the gates were equal we would need the Python interpreter. This commit opts to add a new entrypoint for Rust as working under the constraints that we're only having StandardGates can lead to a more efficient implementation. Additionally, for the preset pass managers we always used the same fixed set of standard gates which don't take parameters to cancel. Since the initial goal of the C API is to expose the preset pass managers this new entrypoint only uses this set which simplifies the checking necessary. This provides an optimized path for the preset pass managers too. For the standalone C function in #14444 we don't need to make the pass configurable. If we decide eventually to expand this it is trivial to add an updated entrypoint, but for the first MVP it's not necessary.

Since this new entrypoint is also more efficient for the use case of the preset pass managers the Python side pass is updated to use it in the default case. This enables the preset pass managers to run the pass faster.

### Details and comments

Related to #14444

mtreinish committed:
Add Python free path to inverse_cancellation

This commit adds a Python free path to the inverse cancellation
transpiler pass. Previously, the only way to run the pass required Python
because the pass handles arbitrary gates to cancel which can include
custom gates from Python. This meant to validate the gates were equal we
would need the Python interpreter. This commit opts to add a new
entrypoint for Rust as working under the constraints that we're only
having StandardGates can lead to a more efficient implementation.
Additionally, for the preset pass managers we always used the same fixed
set of standard gates which don't take parameters to cancel. Since the
initial goal of the C API is to expose the preset pass managers this new
entrypoint only uses this set which simplifies the checking necessary.
This provides an optimized path for the preset pass managers too. For the
standalone C function in #14444 we don't need to make the pass
configurable. If we decide eventually to expand this it is trivial to add
an updated entrypoint, but for the first MVP it's not necessary.

Since this new entrypoint is also more efficient for the use case of the
preset pass managers the Python side pass is updated to use it in the
default case. This enables the preset pass managers to run the pass
faster.

Related to #14444

raynelfss commented on crates/transpiler/src/passes/inverse_cancellation.rs:
```diff
@@ -156,9 +156,113 @@ fn run_on_inverse_pairs(
+            let max_index = gate_cancel_run.len() - 1;
+            for (i, cancel_gate) in gate_cancel_run.iter().enumerate() {
+                let node = &dag[*cancel_gate];
+                let NodeType::Operation(inst) = node else {
+                    unreachable!("Not an op node");
+                };
```
I think that you can use ``NodeType::unwrap_operation()`` here iirc.

```suggestion
                let inst = node.unwrap_operation();
```

mtreinish replied:
I was just following jakelishman's lead https://github.com/Qiskit/qiskit/pull/14317#discussion_r2104637979

raynelfss commented on crates/transpiler/src/passes/inverse_cancellation.rs:
```diff
@@ -156,9 +156,113 @@ fn run_on_inverse_pairs(
+                };
+                let NodeType::Operation(next_inst) = &dag[nodes[i]] else {
+                    unreachable!("Not an op node");
+                };
```
Aren't these the same node??

mtreinish replied:
Oops, good catch. I should add test to cover this explicitly since it clearly wasn't caught

mtreinish replied:
Fixed and tested in: https://github.com/Qiskit/qiskit/pull/14803/commits/f2c10bc7df39e7a5359405d62f919cd5a9511326

raynelfss reviewed (commented):
This looks good so far, I just had a suggestion and a question.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into no-python-inverse-cancellation-for-c-in-the-future

mtreinish committed:
Fix copy paste error in inverse pair path

The inverse pair path was incorrectly checking the same gate twice
instead of the gate and the next one in a run. This was just a copy and
paste error where code was supposed to operate on n and n+1 but it was
doing n twice. This commit fixes this and expands the testing to ensure
we have direct coverage of the new code path instead of relying on the
transpiler tests testing indirectly.

alexanderivrii commented on qiskit/transpiler/passes/optimization/inverse_cancellation.py:
```diff
@@ -21,58 +23,80 @@
-                pair should be inverses of each other.
+                pair should be inverses of each other. If ``None`` a set of standard gates
+                self inverse and inverse gate pairs will be used. The current default list
+                of self inverse gates are:
```
```suggestion
                pair should be inverses of each other. If ``None`` a default list of
                self-inverse gates and a default list of inverse gate pairs will be used.
                The current default list of self-inverse gates is:
```

alexanderivrii commented on qiskit/transpiler/passes/optimization/inverse_cancellation.py:
```diff
@@ -21,58 +23,80 @@
+                  * :class:`.HGate`
+                  * :class:`.SwapGate`
+
+                and the inverse gate pairs are:
```
```suggestion
                and the default list of inverse gate pairs is:
```

alexanderivrii commented on releasenotes/notes/inverse_cancellation_no_python_path-6b0f254a649d2af2.yaml:
```diff
@@ -0,0 +1,25 @@
+    Added a default set of inverse gates for the :class:`.InverseCancellation`
+    transpiler pass. Previously, an explicit list of gates or gate pairs to
+    cancel was a required argument for the constructor of the pass object.
+    Now this list is optional and if no list is provided the self inverse gates:
```
```suggestion
    Now this list is optional and if no list is provided the self inverse gates are:
```

Should we also stress that not providing the optional list is now the default?

mtreinish replied:
It probably wouldn't hurt. I can reword it a bit, or we can save it for the release PR too.

alexanderivrii commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -4368,6 +4368,27 @@ impl DAGCircuit {
 
+impl<'a> DAGCircuit {
+    /// Return an iterator of gate runs with op nodes that match a specified filter function
+    pub fn collect_1q_runs_by<F: Fn(&PackedInstruction) -> bool + 'a>(
```
This is not limited to 1-qubit runs, that is, we could use this function to collect sequence of consecutive CZ-gates, correct?

mtreinish replied:
That's a good point, I'll rename the method `collect_runs_by`.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14803/commits/87134f06b2c2d4eed16b43dafdff2a073f73a71c

alexanderivrii reviewed (commented):
This PR makes a lot of sense, thanks for doing this! From the usability side, having to explicitly specify the lists of self-inverse and inverse pairs in order to call the `InverseCancellation` pass has always gretly annoyed me, the new default argument of `None` fully takes care of this. And of course this improves the performance and enables the planned standalone C function interface.

One could probably remove some code duplication between the existing and the new functions (e.g. between `run_on_self_inverse` and `std_self_inverse`), but I am happy to have the code as is.

While taking a fresh look at the code, I have realized that the `InverseCancellation` pass would not remove CZ(0, 1) followed by CZ(1, 0), or Swap(0, 1) followed by Swap(1, 0), because in general we have to compare (ordered) argument lists and not (unordered) argument sets. Of course, now that we have a flow dealing with standard gates only, we could extend the functionality to handle CZ and Swap gates specially.

We could probably extend a bit the lists of `SELF_INVERSE_GATES_FOR_CANCELLATION` and `INVERSE_PAIRS_FOR_CANCELLATION`, e.g. add `CH` to the first list and `(CS, CSdg)` to the second list (and there are probably other opportunities as well).

I am also wondering if it might be more efficient to call `dag.collect_runs` for all of the known self-inverse gates at once, rather than iteratively doing this for each gate in the SELF_INVERSE list. I am also wondering if in this case it's more efficient to rebuild the DAG vs. iteratively remove nodes from the DAG. I am not suggesting to do this as a part of this PR, this is only a question if you have any experience with other alternative implementations?

Other than that, I only had some nitpicks about function names / docstring comments.

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

mtreinish commented:
> This PR makes a lot of sense, thanks for doing this! From the usability side, having to explicitly specify the lists of self-inverse and inverse pairs in order to call the `InverseCancellation` pass has always gretly annoyed me, the new default argument of `None` fully takes care of this. And of course this improves the performance and enables the planned standalone C function interface.
> 
> One could probably remove some code duplication between the existing and the new functions (e.g. between `run_on_self_inverse` and `std_self_inverse`), but I am happy to have the code as is.

I thought about this but while the logic is very similar the checking we need to do in the standard gate case is significantly simpler. I thought about maybe using a closures to do the actual per gate checking but the logic ended up looking more cluttered than just a parallel path. It also makes the separation between a python mode of execution and the pure-rust a bit more clear which I didn't think was a disadvantage.

> 
> While taking a fresh look at the code, I have realized that the `InverseCancellation` pass would not remove CZ(0, 1) followed by CZ(1, 0), or Swap(0, 1) followed by Swap(1, 0), because in general we have to compare (ordered) argument lists and not (unordered) argument sets. Of course, now that we have a flow dealing with standard gates only, we could extend the functionality to handle CZ and Swap gates specially.
> 
> We could probably extend a bit the lists of `SELF_INVERSE_GATES_FOR_CANCELLATION` and `INVERSE_PAIRS_FOR_CANCELLATION`, e.g. add `CH` to the first list and `(CS, CSdg)` to the second list (and there are probably other opportunities as well).

Feel free to leave a concrete list of suggestions for what to add here, I only copied what was in the preset pass managers but we should try to make the default fairly exhaustive. The incremental cost for adding gates is not that high because we do an upfront check if they're in the circuit before doing the more expensive analysis.

> 
> I am also wondering if it might be more efficient to call `dag.collect_runs` for all of the known self-inverse gates at once, rather than iteratively doing this for each gate in the SELF_INVERSE list. I am also wondering if in this case it's more efficient to rebuild the DAG vs. iteratively remove nodes from the DAG. I am not suggesting to do this as a part of this PR, this is only a question if you have any experience with other alternative implementations?

I thought about this, the problem is this will return runs that can not be simplified. For example if you had the circuit:

```
H - X - H - Y - H - Z
```

If we did it all at once we'd get a run for that sequence but there is not anything to actually cancel. As for rebuilding the dag vs mutating in this case it's hard to know. In general if we remove a large number of gates from the circuit rebuilding would be faster because it has fewer graph operations, but I think we'd need to rethink the algorithm a bit more fundamentally to be able to know up front which is a better approach. Thinking it through there might be a path to do this by rebuilding the dag in reverse, because the `collect_runs()` function does a reverse topological order iteration. But it would be more work and this PR is very targeted to just enable calling the pass from C.
> 
> Other than that, I only had some nitpicks about function names / docstring comments.

mtreinish committed:
Rename collect_1q_runs_by() -> collect_runs_by()

mtreinish committed:
Fix short circuit conditions for inverse pair loop

alexanderivrii commented on crates/transpiler/src/passes/inverse_cancellation.rs:
```diff
@@ -156,9 +156,133 @@ fn run_on_inverse_pairs(
+    StandardGate::Z,
+    StandardGate::H,
+    StandardGate::Swap,
+];
```
If adding more standard self-inverse gates to this list is cheap, we can also include `CH`, `CCX`, `CCZ`, `RCCX`, `CSwap` and `C3X`.

(however the `RC3X` gate is not symmetric, and so should not be included)

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14803/commits/4a201b360bea7dbff2d7c12a3345d31783290c92

alexanderivrii commented on crates/transpiler/src/passes/inverse_cancellation.rs:
```diff
@@ -156,9 +156,133 @@ fn run_on_inverse_pairs(
+    [StandardGate::T, StandardGate::Tdg],
+    [StandardGate::S, StandardGate::Sdg],
+    [StandardGate::SX, StandardGate::SXdg],
+];
```
Here the only extra standard gate pair that we can include is `(CS, CSdg)`.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14803/commits/4a201b360bea7dbff2d7c12a3345d31783290c92

alexanderivrii commented on qiskit/transpiler/passes/optimization/inverse_cancellation.py:
```diff
@@ -21,58 +23,80 @@
 
-    def __init__(self, gates_to_cancel: List[Union[Gate, Tuple[Gate, Gate]]]):
+    def __init__(self, gates_to_cancel: List[Union[Gate, Tuple[Gate, Gate]]] | None = None):
         """Initialize InverseCancellation pass.
```
As we discussed, in the case of user-specified gates (i.e., `gates_to_cancel` is not `None`), from the usability standpoint it might be nice to have an additional boolean argument which specifies whether to also use the default lists of standard self-inverse and inverse-pairs gates. However, now I think that this is orthogonal to the scope of this PR, so it's your call whether you want to leave this for a follow-up.

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/14803/commits/bbc4fa5d5a7a3b58af5c2326d2493a6d3cb5d758

alexanderivrii commented on qiskit/transpiler/passes/optimization/inverse_cancellation.py:
```diff
@@ -21,58 +23,80 @@
                 list is either a single gate or a pair of gates. If a single gate, then
                 it should be self-inverse. If a pair of gates, then the gates in the
-                pair should be inverses of each other.
+                pair should be inverses of each other. If ``None`` a default list of
+                self-inverse gates and a default list of inverse gate pairs will be used.
+                The current default list of self-inverse gates is:
```
Looking at the pre-existing description of this pass, I don't think we explicitly say anywhere that parameterized gates are not allowed. Should we stress this somewhere?

For fun: the Python code can actually be used for canceling parameterized gates with fixed floating-point values of the parameters:
```
qc = QuantumCircuit(5)
qc.append(RXGate(0.1), [1])
qc.append(RXGate(-0.1), [1])

qct = InverseCancellation(gates_to_cancel=[(RXGate(0.1), RXGate(-0.1))])(qc)
```

mtreinish replied:
Yeah, it was a surprise to me too there is explicit test coverage for it. I remember discovering it back when we initially ported this pass to rust as well because I was disappointed that we'd need Python to do gate equality checks because of this and we couldn't just match gate names on their own.

alexanderivrii reviewed (commented):
This looks good. My main remaining suggestion is to add a few more standard self-inverse-gates and inverse-pairs to the default lists, see in-code comments. You might need then to also update the description for the Python pass and the release notes.

alexanderivrii commented:
> I thought about this, the problem is this will return runs that can not be simplified. For example if you had the circuit: `H - X - H - Y - H - Z`

I agree. Also the actual simplification would be a bit more complex: `X - H - H - X - H - Y - Y  - Z` could be simplified to `X - X - H -  Z` or even to `H - Z`, though I believe both of these could be done with the complexity linear in the size of the list.

Anyway, this was me asking for your intuition about which of the alternatives would offer better the better performance and not a concrete implementation suggestion.

mtreinish committed:
Expand defaults lists to include more gates

mtreinish committed:
Add option to run defaults in addition to custom cancellation gates

alexanderivrii reviewed (approved):
Thanks for the updates!
